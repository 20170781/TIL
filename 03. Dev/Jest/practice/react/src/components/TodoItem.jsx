import React, { memo } from 'react';

const TodoItem = memo(({ todo, onIncrement, onDecrement, onDelete }) => {
  const handleIncrement = () => {
    onIncrement(todo);
  };

  const handleDecrement = () => {
    onDecrement(todo);
  };

  const handleDelete = () => {
    onDelete(todo);
  };

  return (
    <li className="todo">
      <span className="todo-name" data-testid="todo-name">
        {todo.name}
      </span>
      <span className="todo-count" data-testid="todo-count">
        {todo.count}
      </span>
      <button
        className="todo-button todo-increase"
        title="increase"
        onClick={handleIncrement}
      >
        <i className="fas fa-plus-square"></i>
      </button>
      <button
        className="todo-button todo-decrease"
        title="decrease"
        onClick={handleDecrement}
      >
        <i className="fas fa-minus-square"></i>
      </button>
      <button
        className="todo-button todo-delete"
        title="delete"
        onClick={handleDelete}
      >
        <i className="fas fa-trash"></i>
      </button>
    </li>
  );
});

export default TodoItem;
