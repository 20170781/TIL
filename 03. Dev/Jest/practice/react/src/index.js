import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './app/app';
import '@fortawesome/fontawesome-free/js/all.js';
import TodoPresenter from './logics/todo_presenter';

const todoPresenter = new TodoPresenter([
  { id: 1, name: 'JavaScript', count: 0 },
  { id: 2, name: 'TypeScript', count: 0 },
  { id: 3, name: 'React', count: 0 },
]);

ReactDOM.render(
  <React.StrictMode>
    <App presenter={todoPresenter} />
  </React.StrictMode>,
  document.getElementById('root')
);
