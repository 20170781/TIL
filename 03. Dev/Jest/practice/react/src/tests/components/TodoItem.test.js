import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoItem from '../../components/TodoItem';

describe('TodoItem', () => {
  const todo = { name: 'javascript', count: 1, id: 1 };

  let TodoItemComp;
  let handleIncrement;
  let handleDecrement;
  let handleDelete;

  beforeEach(() => {
    handleIncrement = jest.fn();
    handleDecrement = jest.fn();
    handleDelete = jest.fn();

    TodoItemComp = (
      <TodoItem
        todo={todo}
        onIncrement={handleIncrement}
        onDecrement={handleDecrement}
        onDelete={handleDelete}
      />
    );
  });

  test('snapshot', () => {
    const component = render(TodoItemComp);
    expect(component.container).toMatchSnapshot();
  });

  describe('Button click', () => {
    beforeEach(() => {
      render(TodoItemComp);
    });

    test('calls handleIncrement', () => {
      const button = screen.getByTitle('increase');
      userEvent.click(button);
      expect(handleIncrement).toHaveBeenCalledWith(todo);
    });

    test('calls handleDecrement', () => {
      const button = screen.getByTitle('decrease');
      userEvent.click(button);
      expect(handleDecrement).toHaveBeenCalledWith(todo);
    });

    test('calls handleDelete', () => {
      const button = screen.getByTitle('delete');
      userEvent.click(button);
      expect(handleDelete).toHaveBeenCalledWith(todo);
    });
  });
});
