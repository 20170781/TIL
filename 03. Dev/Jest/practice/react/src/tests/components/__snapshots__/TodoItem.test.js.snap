// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem snapshot 1`] = `
<div>
  <li
    class="todo"
  >
    <span
      class="todo-name"
      data-testid="todo-name"
    >
      javascript
    </span>
    <span
      class="todo-count"
      data-testid="todo-count"
    >
      1
    </span>
    <button
      class="todo-button todo-increase"
      title="increase"
    >
      <i
        class="fas fa-plus-square"
      />
    </button>
    <button
      class="todo-button todo-decrease"
      title="decrease"
    >
      <i
        class="fas fa-minus-square"
      />
    </button>
    <button
      class="todo-button todo-delete"
      title="delete"
    >
      <i
        class="fas fa-trash"
      />
    </button>
  </li>
</div>
`;
